Adding docs for more components and patterns on Storybook

psatyal opened the pull request:
Added:

- Dropdown button
- Dropdown menu
- Toast

Modified:

- Tooltips

psatyal committed:
Added new documentation

Added:

- Dropdown button
- Dropdown menu
- Toast

Modified:

- Tooltips

psatyal committed:
Added docs for text field

psatyal committed:
Merge branch 'dev' into additional-docs-for-storybook

psatyal commented:
@b12f Requesting review but this is just continuous updates to Storybook so we should be able to merge it.

psatyal committed:
Added docs re: section headers

psatyal commented:
Hi @marcalcobe and @b12f. As mentioned earlier in our frontend-UX daily, I've added some additional documentation for:

- Tooltips (this will surely need to be updated once we updated the component too)
- Drop down button
- Dropdown menu
- Toast
- Text field
- Section headers

Also clarified when header dividers should be used (and not used) for Modal dialogues.

Could you review it, please? Some are more thorough than others, but I thought if we don't merge at some point, this PR will go on forever 😃

psatyal committed:
Added the bit about icon baseline

@b12f Does this work?

marcalcobe commented on frontend/src/app/spot/components/text-field/stories/TextField.stories.mdx:
```diff
@@ -46,33 +52,83 @@ export const Template = (args) => ({
+
+> Example:  A search bar with a icon to the left
+
+An icon can also optionally be placed on the right edge. This arrow can provide additional hints on how the text input field will behave; for example, a right arrow might suggest that the value will be immediately submitted, or a down arrow might indicate that the text field will open a drop modal underneath it (a list of assignees, for example).
```
@psatyal,

In the case of down arrow shouldn't we talk about a dropdown icon instead?

marcalcobe commented on frontend/src/app/spot/components/text-field/stories/TextField.stories.mdx:
```diff
@@ -46,33 +52,83 @@ export const Template = (args) => ({
+
+Clicking on a text field will move it to the focused state. 
+
+A text field can have placeholder text to provide additionanl context or information; this text is replaced by actual text that the user inputs, and is shown again only if the field is cleared.
```
@psatyal, do we need to specify colour and typo of this placeholder text?

psatyal replied:
Nope, we've got a "Focus state" story under Styles. And we don't specify the colour there either, because it might eventually depend on theming. Although if we want, we can add it to that.

marcalcobe commented on frontend/src/app/spot/components/text-field/stories/TextField.stories.mdx:
```diff
@@ -46,33 +52,83 @@ export const Template = (args) => ({
+
+The **Hover** element has a darker border.
+
+The **Disabled** state shows all elements (including icons and any placeholder or input text) in grey, and makes any interaction impossible.
```
Should the grey and darker border be specified? Like which colour hex are them?

psatyal replied:
Yep, I added it. But I'm never sure I should mention them explicitly, since we haven't _actually_ implemented our named colours yet in Storybook.

marcalcobe commented on frontend/src/app/spot/components/tooltip/stories/Tooltip.stories.mdx:
```diff
@@ -65,3 +67,23 @@ export const Template = (args) => ({
+
+The minimum width of the tooltip depends on the length of the text, and the maximum width depends on the width of the containing element.
+
+A tooltip can be multi-line if necessary (?) It is never truncated.
```
I think this is a correct assumption. A tooltip might even content a paragraph if needed. Having a truncated extra information text doesn't make to much sense in my opinion.

marcalcobe commented on frontend/src/stories/DropdownMenu.stories.mdx:
```diff
@@ -0,0 +1,32 @@
+
+The title is not interactive. It serves to provide additional context, if necessary.
+
+A menu item is interactive and generally represents an action. 
```
Should we add here that the menu rows might content an optional icon?

psatyal replied:
Good point! Added.

marcalcobe commented on frontend/src/stories/DropdownMenu.stories.mdx:
```diff
@@ -0,0 +1,32 @@
+
+## States
+
+Each menu item has three states: regular, hover and clicked.
```
Is clicked in this case referring to the option of the menu that is already selected in the dropdown button (selected state) or the action of click?

psatyal replied:
The action of click. But in the case of drop down menus, doesn't "selected" correspond to hover?

marcalcobe commented on frontend/src/stories/SectionHeaders.stories.mdx:
```diff
@@ -0,0 +1,50 @@
+**Left icon**
+This icon can be added to provide context. In a files list, for example, the left icon is used to indicate to which storage provider (for example, Nextcloud) that file storage is linked.
+
+In some parts of the application, this icon is also used to indicate open/closed state when the section header functions as an accordian (see *Behaviour*). 
```
I've never seen the term accordian for this behaviour. I think it is more common the term expandable pannel.

psatyal replied:
It's quite common, but I spelt it wrong here (eg. https://www.freecodecamp.org/news/build-an-accordion-menu-using-html-css-and-javascript/).

I could also do "expandable panel" if that's better, although I'm not sure why it's a panel. Thoughts, @b12f?

marcalcobe commented on frontend/src/stories/SectionHeaders.stories.mdx:
```diff
@@ -0,0 +1,50 @@
+*Note: Since this element is always all-caps, the baseline of the font requires the icon to be raised by a few pixels. See* Margins and Spacing *for more details.*
+
+**Right icon**
+The icon can be placed to the right of the text too, if required. This should be used sparingly.
```
Maybe indicate that this icon can be a link if needed.

psatyal replied:
Added.

marcalcobe commented on frontend/src/stories/SectionHeaders.stories.mdx:
```diff
@@ -0,0 +1,50 @@
+
+## Behaviour
+
+The section header behaves either as a regular header (static) or as an accordian, where clicking on the entire section header will expand and collapse the content underneath.
```
Again the word accordian sounds weird to me.

psatyal replied:
See discussion above.

marcalcobe commented on frontend/src/stories/SectionHeaders.stories.mdx:
```diff
@@ -0,0 +1,50 @@
+
+The headers have an 0.5 rem top and bottom margin. 
+
+Because the label is all-caps, the baseline of the font is (visually) slightly higher than it otherwise would have been with mixed case. If the optional icons are used, the CSS of this component automatically pushed them up by a few pixels for better alignment.
```
Do we need to specify how many pixels are pushed?

psatyal replied:
Do we? I think it might be enough to say it's pushed and then specify it in the CSS. But if we do, I'll need input from the devs. @b12f @Kharonus

marcalcobe commented on frontend/src/stories/Toast.stories.mdx:
```diff
@@ -0,0 +1,51 @@
+- A text
+- A close button
+
+There are four types of toasts, each with a specific icon and colour associated with it:
```
Do we need to specify the colours exactly here with HEX?

psatyal replied:
I suppose these don't change with the theming. But again, I tend to avoid specifying absolute colours till we have a system. Let's discuss this tomorrow. Pinging @b12f and @Kharonus for reference.

psatyal replied:
Added named colours as discussed this morning.

marcalcobe commented on frontend/src/stories/Toast.stories.mdx:
```diff
@@ -0,0 +1,51 @@
+
+> Eg. Blue notification toast with "Load new notifications" link.
+
+Toasts do not disappear automatically; they must be manually closed by the user.
```
Does this apply to everywhere in the current code @b12f?

marcalcobe commented on frontend/src/stories/Toast.stories.mdx:
```diff
@@ -0,0 +1,51 @@
+
+##  Behaviour
+
+The toast appears either as result of user action (a work package attribute is modified, for example) or a background action (there are new notifications).
```
We are currently using in context toast also in multiple setting pages. They are neither triggered by a user action or a background action but are more like a contextual information message. For example in the LDAP groups or LDAP authentication admin settings.

Also this rises the question to me if the EE banners are also extended toasts? Do we treat them separately?

psatyal replied:
Yep, good point. I'll add this to our "things to discuss tomorrow on our daily" list.

psatyal committed:
Updates after Marc's feedback

Updated:

- Text Field
- Tooltip
- Dropdown Menu
- Section headers

There are a few few more open questions, which will be handled in a separate commit.

b12f commented on frontend/src/app/spot/components/text-field/stories/TextField.stories.mdx:
```diff
@@ -104,7 +104,7 @@ The **Default** state indicates that a text field is ready to accept text input.
 The **Focus** state has a coloured outline, the same as any focused element in OpenProject.
 
-The **Hover** element has a darker border.
+The **Hover** element has a darker border (Basic/Gray 1: #333333).
```
I'd refrain from using hex codes here, it just means you'll have to update more docs if you ever update the value, and for daily usage both in figma and in code it's irrelevant anyway

psatyal replied:
Removed hex.

b12f commented on frontend/src/app/spot/components/text-field/stories/TextField.stories.mdx:
```diff
@@ -119,7 +119,7 @@ The **Disabled** state shows all elements (including icons and any placeholder o
 </Canvas>
 
-The **Error** state puts a red outline around the text input. (_Note_: This state is only used when the text input is part of a _Form element_, which supplements this red outline with an message error message). The error state is usually temporary; fixing the cause of the error (via asyncronous form validation) usually allows it to revert back to the focus 
+The **Error** state puts a red outline around the text input. (_Note_: This state is only used when the text input is part of a _Form element_, which supplements this red outline with an message error message). The error state is usually temporary; fixing the cause of the error (via asyncronous form validation) usually allows it to revert back to the focus state.
```
can also be synchronous form validation btw, so I'd just call it "via form validation"

psatyal committed:
Minor changes after our daily today

b12f reviewed (approved):
LGTM
